Enhance searchable snapshots to enable a read-only view of older snapshots

kartg opened the pull request:
### Description

This change removes the guardrails around N-1 backward compatibility and uses Lucene's "expert" APIs to read snapshots (Lucene segments) older than N-1 on a best-effort basis. The functionality is gated by an additional feature flag, separate from the searchable snapshots flag. Note that the Lucene integration is rather inefficient because the necessary "expert" Lucene APIs are still package-private.

Signed-off-by: Kartik Ganesh <gkart@amazon.com>

### Issues Resolved
closes https://github.com/opensearch-project/OpenSearch/issues/5451

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

mch2 commented on server/src/main/java/org/opensearch/Version.java:
```diff
@@ -413,6 +416,9 @@ private Version computeMinIndexCompatVersion() {
         } else if (major == 7 || major == 1) {
             return LegacyESVersion.fromId(6000026);
         } else if (major == 2) {
+            if (FeatureFlags.isEnabled(SEARCHABLE_SNAPSHOT_EXTENDED_BWC)) {
```
This also has impact on non SS indices.  Could this be an additional setting within SS?  What if this was explicitly set for SS in your update `readAnySegmentInfos ` below to call `minimumIndexCompatibilityVersion` ?

mch2 commented on server/src/main/java/org/opensearch/index/engine/ReadOnlyEngine.java:
```diff
@@ -115,10 +130,12 @@ public ReadOnlyEngine(
         Function<DirectoryReader, DirectoryReader> readerWrapperFunction,
-        boolean requireCompleteHistory
+        boolean requireCompleteHistory,
+        boolean allowOldIndices
```
Rather than overloading here with the allow flag, could you put this as extra into indexSettings?  Then you could tie it as a special setting for SS only indices when parsing settings?

mch2 commented on server/src/main/java/org/opensearch/index/store/Store.java:
```diff
@@ -221,15 +223,22 @@ public Directory directory() {
     public SegmentInfos readLastCommittedSegmentsInfo() throws IOException {
```
Not quite following why this update is required (and the additional `readAnySegmentsInfo `) when it doesn't get invoked from ReadOnlyEngine?

mch2 commented on server/src/main/java/org/opensearch/index/shard/IndexShard.java:
```diff
@@ -2054,6 +2061,10 @@ private boolean assertSequenceNumbersInCommit() throws IOException {
         return true;
     }
 
+    private boolean isRemoteSnapshotIndex() {
```
This helper function could be set on IndexSettings earlier.

mch2 commented on server/src/main/java/org/opensearch/index/shard/IndexShard.java:
```diff
@@ -2039,7 +2040,13 @@ private void innerOpenEngineAndTranslog(LongSupplier globalCheckpointSupplier) t
+            // Inefficient method to support reading old Lucene indexes
+            userData = Lucene.readAnySegmentInfos(store.directory()).getUserData();
+        } else {
+            userData = Lucene.readSegmentInfos(store.directory()).getUserData();
```
why is the fallback here different & not `SegmentInfos.readLatestCommit(store.directory()).getUserData();` ?

mch2 commented on server/src/main/java/org/opensearch/common/lucene/Lucene.java:
```diff
@@ -162,6 +163,23 @@ public static SegmentInfos readSegmentInfos(Directory directory) throws IOExcept
+     * the latest index commit is opened via {@link DirectoryReader} by including a minimum supported
+     * Lucene major version - this is based on the minimum index compatiblity version of OpenSearch.
+     */
+    public static SegmentInfos readAnySegmentInfos(Directory directory) throws IOException {
```
This is set up to be a utility method, but the ability to read "any" infos depends on the minimumIndexCompatibilityVersion.  Could we isolate this logic to SS only inside of ReadOnlyEngine?

kartg commented on server/src/main/java/org/opensearch/cluster/metadata/IndexMetadata.java:
```diff
@@ -1825,10 +1825,13 @@ public static IndexMetadata fromXContent(XContentParser parser) throws IOExcepti
                 assert mappingVersion : "mapping version should be present for indices";
             }
-            if (Assertions.ENABLED) {
+            if (Assertions.ENABLED && Version.indexCreated(builder.settings).onOrAfter(legacyVersion)) {
```
Oops, good spot. Fixed.

gregschohn commented on server/src/test/java/org/opensearch/index/store/StoreTests.java:
```diff
@@ -1257,6 +1259,63 @@ public void testSegmentReplicationDiff() {
+        final ShardId shardId = new ShardId("index", "_na_", 1);
+        ArrayList<Settings> testSettingsList = new ArrayList<>();
+        // non-snapshot use-case
+        testSettingsList.add(
```
Curious - is there a notion of parameterized tests, like what JUnit provides.

gregschohn reviewed (commented):
Please add a comment - maybe where you use the testIndex file or a readme file alongside it - to show how you generated that file.  A set of commands and specific versions would be great.

kartg commented:
> Please add a comment - maybe where you use the testIndex file or a readme file alongside it - to show how you generated that file. A set of commands and specific versions would be great.

Good callout, @gregschohn. I've added a README file in the latest commit 😄

gregschohn commented on server/src/main/java/org/opensearch/index/shard/IndexShard.java:
```diff
@@ -2039,7 +2039,16 @@ private void innerOpenEngineAndTranslog(LongSupplier globalCheckpointSupplier) t
     private boolean assertSequenceNumbersInCommit() throws IOException {
-        final Map<String, String> userData = SegmentInfos.readLatestCommit(store.directory()).getUserData();
+        final Map<String, String> userData;
+        if (indexSettings.isRemoteSnapshot() && indexSettings.getExtendedCompatibilitySnapshotVersion() != null) {
+            // Inefficient method to support reading old Lucene indexes
+            userData = Lucene.readSegmentInfosExtendedCompatibility(
+                store.directory(),
+                indexSettings.getExtendedCompatibilitySnapshotVersion()
+            ).getUserData();
+        } else {
+            userData = SegmentInfos.readLatestCommit(store.directory()).getUserData();
+        }
```
nit: The code would read a bit better if this were abstracted into a separate function

gregschohn commented on server/src/main/java/org/opensearch/index/store/Store.java:
```diff
@@ -245,7 +251,27 @@ private static SegmentInfos readSegmentsInfo(IndexCommit commit, Directory direc
+     *
+     * @throws IOException if the index is corrupted or the segments file is not present
+     */
+    private static SegmentInfos readSegmentInfosExtendedCompatbility(Directory directory, org.opensearch.Version minimumVersion)
```
spelling

kartg replied:
Man, i cannot seem to spell "compatibility" correctly to save my life 😄

kartg commented on server/src/main/java/org/opensearch/common/lucene/Lucene.java:
```diff
@@ -162,6 +163,27 @@ public static SegmentInfos readSegmentInfos(Directory directory) throws IOExcept
+        List<IndexCommit> indexCommits = DirectoryReader.listCommits(directory);
+        IndexCommit latestCommit = indexCommits.get(indexCommits.size() - 1);
+        final int minSupportedLuceneMajor = minimumVersion.minimumIndexCompatibilityVersion().luceneVersion.major;
+        StandardDirectoryReader reader = (StandardDirectoryReader) DirectoryReader.open(latestCommit, minSupportedLuceneMajor, null);
+        // cache the SegmentInfos object before closing the reader to prevent any open file handles
+        SegmentInfos si = reader.getSegmentInfos();
```
Wrote a (rather basic) unit test to check for this. I'd argue that the two implementations are actually coupled but it's just difficult to see given the workaround implementation in this method.

The code in `readSegmentInfos` internally invokes the "expert" API, but sets the minimum-compatible version to N-1:

https://github.com/apache/lucene/blob/3e8ef57e3fa1a015fc4ae22153660b21626886c4/lucene/core/src/java/org/apache/lucene/index/SegmentInfos.java#L517

Ideally, the `readSegmentInfosExtendedCompatibility` method would simply invoke this expert API and override the minimum-compatible version. However, since that method signature is package-private in Lucene, the round-a-bout way to access it is through `DirectoryReader.listCommits` which uses it internally - 

https://github.com/apache/lucene/blob/3e8ef57e3fa1a015fc4ae22153660b21626886c4/lucene/core/src/java/org/apache/lucene/index/DirectoryReader.java#L267

and then enforce the minimum-compatible version via the `DirectoryReader`.

That said, I take your point that the two code flows aren't identical and could change underneath us, so the unit test is valuable.

reta commented on server/src/main/java/org/opensearch/common/lucene/Lucene.java:
```diff
@@ -162,6 +163,27 @@ public static SegmentInfos readSegmentInfos(Directory directory) throws IOExcept
+        final int minSupportedLuceneMajor = minimumVersion.minimumIndexCompatibilityVersion().luceneVersion.major;
+        StandardDirectoryReader reader = (StandardDirectoryReader) DirectoryReader.open(latestCommit, minSupportedLuceneMajor, null);
+        // cache the SegmentInfos object before closing the reader to prevent any open file handles
+        SegmentInfos si = reader.getSegmentInfos();
```
```suggestion
   try (StandardDirectoryReader reader = (StandardDirectoryReader) DirectoryReader.open(latestCommit, minSupportedLuceneMajor, null)) {
        return reader.getSegmentInfos();
   }
```

reta commented on server/src/main/java/org/opensearch/index/IndexSettings.java:
```diff
@@ -747,6 +756,13 @@ public IndexSettings(final IndexMetadata indexMetadata, final Settings nodeSetti
+        if (isRemoteSnapshot) {
+            Integer minVersion = settings.getAsInt(SEARCHABLE_SNAPSHOT_MINIMUM_VERSION.getKey(), Version.V_EMPTY_ID);
+            if (minVersion > Version.V_EMPTY_ID) {
+                extendedCompatibilitySnapshotVersion = LegacyESVersion.fromId(minVersion);
```
Question here please: so for now `extendedCompatibilitySnapshotVersion` is treated as legacy Elasticsearch version, but could it be eventually an OpenSearch version (let say in 3.x it could be set to 1.x)?

kartg replied:
That's correct! And I'd very much like to improve this code to support that. However, doing so would require heavy refactoring in the `Version` and `LegacyESVersion` classes to support better comparisons between the two. So for now, I've limited this implementation to legacy versions.

reta commented on server/src/main/java/org/opensearch/index/engine/ReadOnlyEngine.java:
```diff
@@ -93,9 +93,24 @@ public class ReadOnlyEngine extends Engine {
+        Function<DirectoryReader, DirectoryReader> readerWrapperFunction,
+        boolean requireCompleteHistory
+    ) {
+        this(config, seqNoStats, translogStats, obtainLock, readerWrapperFunction, requireCompleteHistory, null);
```
```suggestion
        this(config, seqNoStats, translogStats, obtainLock, readerWrapperFunction, requireCompleteHistory, Version.MIN_SUPPORTED_MAJOR);
```

reta replied:
I think usage of `null` here is very confusing: the `minimumSupportedVersion` is not going anywhere, it is `Version.MIN_SUPPORTED_MAJOR`

andrross commented on server/src/main/java/org/opensearch/index/IndexSettings.java:
```diff
@@ -575,6 +575,13 @@ public final class IndexSettings {
         Property.InternalIndex
     );
 
+    public static final Setting<Integer> SEARCHABLE_SNAPSHOT_MINIMUM_VERSION = Setting.intSetting(
```
Why is this configurable? Isn't there an inherent limit as to what the software can actually support?

kartg replied:
Yes, there is an inherent limit, and modeling this as a setting makes it seem configurable when it really isn't. The purpose was to embed the minimum version value into index settings (as per [this comment](https://github.com/opensearch-project/OpenSearch/pull/5429#discussion_r1037656013)) so that this logic was limited to searchable snapshots ([code](https://github.com/kartg/OpenSearch/blob/94a9ad0e9787b141eb8047661c56346f61ee56ba/server/src/main/java/org/opensearch/snapshots/RestoreService.java#L1256-L1270)). As shown in that code snippet, the actual value of the setting is hard-coded.

Any suggestions on how I could make this less ugly?

andrross replied:
The _version_ is not a property of the index itself; it's a property of the software, which can potentially change without the data the makes up the index changing. So I don't think it makes sense as an index setting. Can you change this to pass the hard-coded version value when you construct the `ReadOnlyEngine` instance? That is already inside a "if remote snapshot" block so that seems like it's just moving the hard-coded behavior and gets rid of the index setting.

andrross commented on server/src/main/java/org/opensearch/common/util/FeatureFlags.java:
```diff
@@ -37,6 +37,13 @@ public class FeatureFlags {
+     * Gates the ability for Searchable Snapshots to read snapshots that are older than the
+     * guaranteed backward compatibility for OpenSearch (one prior major version) on a best effort basis.
+     */
+    public static final String SEARCHABLE_SNAPSHOT_EXTENDED_COMPATIBILITY =
```
This might be a bad idea, but can we consider not adding another feature flag for this? If we need to promote searchable snapshots out of experimental without the extended compatibility support, we can put a new feature flag in at that time.

kartg replied:
I'm ok with doing this. I took the more conservative route to protect against a contributor removing the searchable snapshots feature flag before the backwards compatibility changes were sufficiently tested.

andrross commented on server/src/main/java/org/opensearch/common/lucene/Lucene.java:
```diff
@@ -162,6 +163,25 @@ public static SegmentInfos readSegmentInfos(Directory directory) throws IOExcept
+    /**
+     * A variant of {@link #readSegmentInfos(Directory)} that supports reading indices written by
+     * older major versions of Lucene. The underlying implementation is a workaround since the
+     * "expert" readLatestCommit API is currently package-private in Lucene. First, all commits in
```
I have a couple questions that might be worth answering in this comment:

- Why is `org.apache.lucene.index.SegmentInfos.readLatestCommit(Directory, int)` package private?
- How does this implementation differ from the built-in (package-private) Lucene version?
- Is there a plan to swap this implementation out for the built-in Lucene version at some point?

kartg replied:
1. In a word - 🤷 The [Lucene commit that introduced it](https://github.com/apache/lucene/commit/c1ae6dc07c9a988533cbe7176bdeb49e2fca1d9c) kept it package-private, with the only public access to it through `DirectoryReader.open` which is the workaround used in this method implementation.

---

2. The expert API in Lucene fetches _only_ the latest `segments_N` file and reads the `SegmentInfos` object from it. The implementation here does the following:
* Under `DirectoryReader.listCommits`, It reads the `SegmentInfos` object for the latest `segments_N` file _as well as for all older_ `segments_N` files. These are returned as a sorted list of `IndexCommit` objects which are created from the corresponding `SegmentInfos`but do not hold a reference to them.
* The latest `IndexCommit` object is then used to open a `DirectoryReader`. Underneath, the `IndexCommit` is transformed to a `SegmentInfos` object by the same method as the expert API. This `SegmentInfos` object is the one that is finally returned from this method.
  
--- 

5. I don't have any information on when Lucene plans on changing the access modifier, so I don't have a plan 😞 though I certainly _want_ to swap this out for the Lucene version.

kartg replied:
I should also note that the two code paths described in point 2 above have an associated test case in LuceneTests - `testReadSegmentInfosExtendedCompatibilityBaseCase`

andrross commented on server/src/main/java/org/opensearch/index/IndexSettings.java:
```diff
@@ -747,6 +756,13 @@ public IndexSettings(final IndexMetadata indexMetadata, final Settings nodeSetti
         isRemoteStoreEnabled = settings.getAsBoolean(IndexMetadata.SETTING_REMOTE_STORE_ENABLED, false);
         remoteStoreRepository = settings.get(IndexMetadata.SETTING_REMOTE_STORE_REPOSITORY);
         isRemoteTranslogStoreEnabled = settings.getAsBoolean(IndexMetadata.SETTING_REMOTE_TRANSLOG_STORE_ENABLED, false);
+        isRemoteSnapshot = IndexModule.Type.REMOTE_SNAPSHOT.match(this);
```
I know this works now, but leaking the `this` pointer from a Java constructor is usually a really bad idea. `IndexModule.Type.REMOTE_SNAPSHOT.match` gets a reference to a partially constructed object, which can lead to nasty failures (e.g. final fields can be observed as being unset). I'd refactor this to get the boolean value you want from the `settings` variable.

kartg replied:
Updated - I was trying to avoid repeating the logic under [IndexModule.match](https://github.com/opensearch-project/OpenSearch/blob/ea675c34b34d3af0efdd650861648a4ef7275881/server/src/main/java/org/opensearch/index/IndexModule.java#L465-L468) 😄

andrross replied:
Feel free to make a new `IndexModule.Type#match(Settings)` overloaded version

andrross commented on server/src/test/resources/indices/bwc/testIndex-6.3.0.md:
```diff
@@ -0,0 +1,57 @@
+# README for _testIndex-6.3.0.zip_
```
Just an organizational nitpick, but would it make sense to created an `es-6.3.0` (or similarly named) subfolder here? Then this file could just be README.md and refer to all files in its folder, though there is only one at this point.

Also I'd put "es" or something similar in the file name. This has the potential to be around for a long long time and will eventually conflict with an OpenSearch 6.x version.

kartg commented on server/src/main/java/org/opensearch/index/store/remote/directory/RemoteSnapshotDirectory.java:
```diff
@@ -35,6 +37,9 @@
  */
 public final class RemoteSnapshotDirectory extends Directory {
+
+    public static final Version SEARCHABLE_SNAPSHOT_EXTENDED_COMPATIBILITY_MINIMUM_VERSION = LegacyESVersion.fromId(6000099);
```
Happy to hear feedback on a better place to stick this. Ideally, this should be modeled as a part of the `FeatureFlag` but that would require a broader change to allow feature "flags" to support values beyond just `boolean`.

andrross replied:
I think it is fine not being a part of the feature flag. This place seems as good as any to me as far as where to put it now.

andrross commented on server/src/test/java/org/opensearch/index/store/StoreTests.java:
```diff
@@ -1257,6 +1263,54 @@ public void testSegmentReplicationDiff() {
+        final ShardId shardId = new ShardId("index", "_na_", 1);
+        Store store = null;
+        try {
+            AccessController.doPrivileged(
+                (PrivilegedAction<String>) () -> System.setProperty(FeatureFlags.SEARCHABLE_SNAPSHOT_EXTENDED_COMPATIBILITY, "true")
+            );
```
Can you make this a bit more concise by creating an AutoCloseable helper class? The syntax would then look like:

```
try (FeatureFlagSetter setter = FeatureFlagSetter.set(FeatureFlags.SEARCHABLE_SNAPSHOT_EXTENDED_COMPATIBILITY)) {
    ... test impl ...
}
```

The class would be:


```
public final class FeatureFlagSetter implements AutoCloseable {
    private final String flag;

    private FeatureFlagSetter(String flag) {
        this.flag = flag;
    }

    public static FeatureFlagSetter set(String flag) {
        AccessController.doPrivileged((PrivilegedAction<String>) () -> System.setProperty(flag, "true"));
        return new FeatureFlagSetter(flag);
    }

    @Override
    public void close() throws Exception {
        AccessController.doPrivileged((PrivilegedAction<String>) () -> System.clearProperty(flag));
    }
}
```

mch2 commented on server/src/main/java/org/opensearch/index/engine/ReadOnlyEngine.java:
```diff
@@ -223,7 +253,17 @@ protected final OpenSearchDirectoryReader wrapReader(
+    }
+
+    private boolean isExtendedCompatibility() {
+        return this.minimumSupportedVersion.before(Version.CURRENT.minimumIndexCompatibilityVersion());
```
`getExtendedCompatibilitySnapshotVersion` will return null if the FF is not enabled, which is what is invoked when the new ctor is called from IndexService.  Would this not throw a NPE when FF is disabled?

kartg replied:
Good spot. The current code manages to avoid this case, but doesn't explicitly protect against it. I'll fix this and add a related test case as well

kartg replied:
Implicitly protected against an NPE by adding defaulting behavior in `IndexSettings` when it computes `extendedCompatibilitySnapshotVersion`

mch2 commented on server/src/main/java/org/opensearch/index/engine/ReadOnlyEngine.java:
```diff
@@ -108,17 +131,20 @@ public class ReadOnlyEngine extends Engine {
         Function<DirectoryReader, DirectoryReader> readerWrapperFunction,
-        boolean requireCompleteHistory
+        boolean requireCompleteHistory,
+        Version minimumSupportedVersion
```
I don't think this overloading is required, you can read the `extendedCompatibilitySnapshotVersion` directly from config. `config().getIndexSettings().getExtendedCompatibilitySnapshotVersion()`

andrross commented:
Looks like I clicked approve a bit too soon...there are some test failures :)

kartg commented:
> Looks like I clicked approve a bit too soon...there are some test failures :)

@andrross oops, my mistake - i changed the defaulting behavior in `IndexSettings` but forgot to update the related test in `IndexSettingsTests`. Fixed now, and rebased to resolve the `CHANGELOG` merge conflict.

---

Also the `unstable` result above seems to be coming from a flaky test failure - #3603
